token-swap/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+                let swap: &TokenSwap = unpack(input)?;
+                Self::Init(*swap)
+            }
+            _ => return Err(ProgramError::InvalidAccountData),
+        })
+    }
+
+    pub fn serialize(self: &Self, output: &mut [u8]) -> ProgramResult {
+        if output.len() < size_of::<u8>() {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        match self {
+            Self::Unallocated => output[0] = 0,
+            Self::Init(swap) => {
+                if output.len() < size_of::<u8>() + size_of::<TokenSwap>() {
+                    return Err(ProgramError::InvalidAccountData);
+                }
+                output[0] = 1;
+                #[allow(clippy::cast_ptr_alignment)]
+                let value = unsafe { &mut *(&mut output[1] as *mut u8 as *mut TokenSwap) };
+                *value = *swap;
+            }
+        }
+        Ok(())
+    }
+
+    fn token_swap(&self) -> Result<TokenSwap, ProgramError> {
+        if let State::Init(swap) = &self {
+            Ok(*swap)
+        } else {
+            Err(Error::InvalidState.into())
+        }
+    }
+
+    pub fn token_account_deserialize(
+        info: &AccountInfo,
+    ) -> Result<spl_token::state::Account, Error> {
+        if let Some(spl_token::state::State::Account(account)) =
+            spl_token::state::State::deserialize(&info.data.borrow()).ok()
+        {
+            Ok(account)
+        } else {
+            Err(Error::ExpectedAccount)
+        }
+    }
+
+    pub fn token_deserialize(info: &AccountInfo) -> Result<spl_token::state::Token, Error> {
+        if let Some(spl_token::state::State::Token(token)) =
+            spl_token::state::State::deserialize(&info.data.borrow()).ok()
+        {
+            Ok(token)
+        } else {
+            Err(Error::ExpectedToken)
+        }
+    }
+
+    pub fn authority_id(program_id: &Pubkey, my_info: &Pubkey) -> Result<Pubkey, Error> {
+        Pubkey::create_program_address(&[&my_info.to_string()[..32]], program_id)
+            .or(Err(Error::InvalidProgramAddress))
+    }
+    pub fn token_burn(
+        accounts: &[AccountInfo],
+        swap: &Pubkey,
+        authority: &Pubkey,
+        token: &Pubkey,
+        burn_account: &Pubkey,
+        amount: u64,
+    ) -> Result<(), ProgramError> {
+        let swap_string = swap.to_string();
+        let signers = &[&[&swap_string[..32]][..]];
+        let ix = spl_token::instruction::burn(
+            &Pubkey::default(),
+            authority,
+            burn_account,
+            token,
+            None,
+            amount,
+        )?;
+        invoke_signed(&ix, accounts, signers)
+    }
+
+    pub fn token_mint_to(
+        accounts: &[AccountInfo],
+        swap: &Pubkey,
+        authority: &Pubkey,
+        token: &Pubkey,
+        destination: &Pubkey,
+        amount: u64,
+    ) -> Result<(), ProgramError> {
+        let swap_string = swap.to_string();
+        let signers = &[&[&swap_string[..32]][..]];
```

Comment:
This is unintuitive and clutters the code. What about making this a function in the SDK taking n seed strings to output signers?

Reply:
Generally the whole aspect of signer seeds is undocumented and rather unintuitive. I see why it's there and what it does but it required me to read the actual syscall implementation.